sound/soc/sof/loader.c

```diff
@@ -358,6 +360,94 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module *fm_entry;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	if (fw->size < hdr->len + SOF_MAN4_FW_HDR_OFFSET) {
+		dev_err(sdev->dev, "Invalid hdr->len %d\n", hdr->len);
+		return -EINVAL;
+	}
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
+							 SOF_MAN4_FW_HDR_OFFSET);
+
+	if (fw->size < ((void *)fw_header - (void *)fw->data) + fw_header->len) {
+		dev_err(sdev->dev, "Invalid fw_header->len %d\n", fw_header->len);
+		return -EINVAL;
+	}
+
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module *)((void *)fw_header + fw_header->len);
+
+	if (fw->size < ((void *)fm_entry - (void *)fw->data) +
+			fw_header->num_module_entries * sizeof(*fm_entry)) {
```

Comment:
It might be again probably just me, but I find this more digestible:
```c
if (fw->size < (hdr->len + SOF_MAN4_FW_HDR_OFFSET + fw_header->len +
		fw_header->num_module_entries * sizeof(*fm_entry)) {
```
and do the `fm_entry = ...` after this check?

Reply:
Does each fm_entry must have a matching fm_config? If so then that should be also incorporated into this check?

Reply:
> Does each fm_entry must have a matching fm_config? If so then that should be also incorporated into this check?

No, `fm_entry->cfg_count` can be 0 which means there is no fm_config for this fm_entry.